### PR TITLE
Add Google Pay branded button

### LINKED ISSUE #1071: Stripe Payment Request button using incorrect Google Pay brand asset
p7bbVw-3af-p2

https://href.li/?https://wordpress.org/support/topic/change-text-on-google-and-apple-pay-button/#post-11539622

Related: https://github.com/woocommerce/woocommerce-gateway-stripe/issues/751

### PR DESCRIPTION
**Pull request by luizreis**
Fixes #1071.

#### Changes proposed in this Pull Request:

- Introduce `Branded` button type
- Add a branded button for Google Pay, to be displayed in Chrome
 
#### Testing instructions:

**Admin settings:**

1. Go to WP Admin > WooCommerce > Settings > Payments > Stripe.
2. Select the `Branded` type for the Payment Request button. The Payment Request Branded Button Type option should appear below. Switching type to something else should hide it.
3. Select a theme, enter a height, and select a branded button type for the Payment Request button.
4. Save changes.

**Custom payment button:**
It should be tested with ngrok enabled and different browsers.
It should be tested with different admin settings ( `branded type: long or short`, and `theme: dark or light`).
It should be tested with different locales.
It should be tested for different products: simple, digital, variable products.

**Only Chrome browsers should display a Google Pay branded button. Safari should display the Apple Pay button and other browsers should display the Stripe payment button.**

1. Go to a product page and make sure the branded button is rendered according to the admin settings.
2. Test the button behavior with different counts and product variations.
3. Go to the cart page.
4. Test if the branded button is rendered and works as expected.

### CODE REVIEW

File: assets/js/stripe-payment-request.js [hunk 1]
```diff
@@ -467,6 +486,32 @@ jQuery( function( $ ) {
 			return prButton && 'function' === typeof prButton.data && prButton.data( 'isCustom' );
 		},
 
+		isBrandedPaymentRequestButton: function ( prButton ) {
+			return prButton && 'function' === typeof prButton.data && prButton.data( 'isBranded' );
+		},
+
+		shouldUseGooglePayBrand: function () {
+			return window.navigator.userAgent.match(/Chrome\/([0-9]+)\./i) && 'Google Inc.' == window.navigator.vendor;
+		},
```
Comment (vbelolapotkov) [inline assets/js/stripe-payment-request.js:495]: I'm wondering is it better and safer to move that detection and button rendering to the server like we do with custom button at the moment? And leverage `css_selector` param value to hook into button button rendered by the site. Personally I feel better if this kind of rendering and logic resides on the site backend.

What do you think?

**Comment by vbelolapotkov** [comment]
Extra note, the rebase and assets rebuild is required to resolve conflicts 🙂

File: assets/css/stripe-styles.scss [hunk 2]
```diff
@@ -62,3 +62,57 @@
 .woocommerce-SavedPaymentMethods-token .stripe-source-errors .woocommerce-error { margin-top: 1em; margin-bottom: 2em; }
 
 #wc-stripe-custom-button { display: block; width: 100%; }
+
```
[the remark this answers is not in the document]
Comment (vbelolapotkov) [reply assets/css/stripe-styles.scss:65]: 👍

File: assets/js/stripe-payment-request.js [hunk 3]
```diff
@@ -444,6 +444,25 @@ jQuery( function( $ ) {
...
+					);
+					button.css( 'height', wc_stripe_payment_request_params.button.height + 'px' );
+					// Add flag to be sure that created button is branded rather than fallback element.
+					button.data( 'isBranded', true );
+					return button;
+				} else {
+					// Not implemented branded buttons default to Stripe's button
+					// Apple Pay buttons can also fall back to Stripe's button, as it's already branded
+					// Set button type to default to avoid issues with Stripe
+					wc_stripe_payment_request_params.button.type = 'default';
```
[the remark this answers is not in the document]
Comment (vbelolapotkov) [reply assets/js/stripe-payment-request.js:457]: I haven't done anything specific. Ok, let's leave as is for now.

File: assets/css/stripe-styles.scss [hunk 4]
```diff
@@ -62,3 +62,57 @@
...
+    background-origin: content-box;
+    background-position: center center;
+    background-repeat: no-repeat;
+    background-size: contain;
+    border: 0px;
+    border-radius: 4px;
+    box-shadow: rgba(60, 64, 67, 0.3) 0px 1px 1px 0px, rgba(60, 64, 67, 0.15) 0px 1px 3px 1px;
+    cursor: pointer;
+    height: 40px;
+    min-height: 40px;
```
[the remark this answers is not in the document]
Comment (vbelolapotkov) [reply assets/css/stripe-styles.scss]: Probably we should ignore the height for the branded buttons, but let's keep it as is for now and improve later if required.

File: assets/css/stripe-styles.scss [hunk 5]
```diff
@@ -62,3 +62,57 @@
 .woocommerce-SavedPaymentMethods-token .stripe-source-errors .woocommerce-error { margin-top: 1em; margin-bottom: 2em; }
 
 #wc-stripe-custom-button { display: block; width: 100%; }
+
+.gpay-button {
+    background-origin: content-box;
+    background-position: center center;
+    background-repeat: no-repeat;
+    background-size: contain;
```
[the remark this answers is not in the document]
Comment (vbelolapotkov) [reply assets/css/stripe-styles.scss]: 👍 Totally makes sense. Let's keep it.

File: assets/css/stripe-styles.scss [hunk 6]
```diff
@@ -62,3 +62,57 @@
...
+		&:focus {
+			box-shadow: #e8e8e8 0 1px 1px 0, #e8e8e8 0 1px 3px;
+			outline: 0;
+		}
+	}
+
+	&.dark {
+		background-color: #000;
+		box-shadow: none;
+		padding: 12px 24px 10px;
```
[the remark this answers is not in the document]
Comment (vbelolapotkov) [reply assets/css/stripe-styles.scss:99]: 🙂

File: includes/admin/stripe-settings.php [hunk 7]
```diff
@@ -157,6 +158,18 @@
...
+		'payment_request_button_branded_type' => array(
+			'title'       => __( 'Payment Request Branded Button Type', 'woocommerce-gateway-stripe' ),
+			'label'       => __( 'Branded Button Type', 'woocommerce-gateway-stripe' ),
+			'type'        => 'select',
+			'description' => __( 'Select the type of button you would like to show.', 'woocommerce-gateway-stripe' ),
+			'default'     => 'long',
+			'desc_tip'    => true,
+			'options'     => array(
+				'short' => __( 'Short', 'woocommerce-gateway-stripe' ),
+				'long'  => __( 'Long', 'woocommerce-gateway-stripe' ),
```
[the remark this answers is not in the document]
Comment (vbelolapotkov) [reply includes/admin/stripe-settings.php]: Fantastic, thank you!

File: assets/js/stripe-payment-request.js [hunk 8]
```diff
@@ -444,6 +444,25 @@ jQuery( function( $ ) {
...
+			if ( wc_stripe_payment_request_params.button.is_branded ) {
+				if ( wc_stripe_payment_request.shouldUseGooglePayBrand() ) {
+					button = wc_stripe_payment_request.createGooglePayButton(
+						wc_stripe_payment_request_params.button.theme,
+						wc_stripe_payment_request_params.button.branded_type,
+						wc_stripe_payment_request_params.button.locale
+					);
+					button.css( 'height', wc_stripe_payment_request_params.button.height + 'px' );
+					// Add flag to be sure that created button is branded rather than fallback element.
+					button.data( 'isBranded', true );
```
[the remark this answers is not in the document]
Comment (vbelolapotkov) [reply assets/js/stripe-payment-request.js]: Looks good! Thank you!

File: assets/js/stripe-payment-request.js (hunk 1 above)
Comment (vbelolapotkov) [reply assets/js/stripe-payment-request.js:490]: > I think rendering it in the JS is more aligned to what Stripe does when rendering their button.

Stripe is the loaded library from 3rd party resource. Don't think they have the other choice :) 
But their https request is not rendering any markup. 

Anyway, let's keep it on the client for now, but let's create an `Engineering` issue for that as well. IMO worth asking for other opinions here.

Comment (luizreis) [reply assets/js/stripe-payment-request.js:518]: Agreed! I've created #1212.

**Pull request merged**